c++/src/Timezone.cc

```diff
@@ -691,12 +683,68 @@ namespace orc {
     try {
       std::unique_ptr<InputStream> file = readFile(filename);
       size_t size = static_cast<size_t>(file->getLength());
-      std::vector<unsigned char> buffer(size);
+      buffer.resize(size);
       file->read(&buffer[0], size, 0);
-      timezoneCache[filename] = std::make_shared<TimezoneImpl>(filename, buffer);
     } catch (ParseError& err) {
       throw TimezoneError(err.what());
     }
+    return buffer;
+  }
+
+  class LazyTimezone : public Timezone {
+   private:
+    std::string filename_;
+    mutable std::unique_ptr<TimezoneImpl> impl_;
+    mutable std::atomic<bool> initialized_;
+
+    TimezoneImpl* getImpl() const {
+      if (!initialized_) {
+        std::lock_guard<std::mutex> timezone_lock(timezone_mutex);
```

Comment:
`timezone_mutex` is used to guard `TimezoneImpl::timezoneCache`. Should we use a much finer-grained mutex which is per LazyTimezone instance?